src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBindingHelper.cs

```diff
@@ -654,28 +653,31 @@ internal static string GetPropertyName(Expression expression)
                 throw new ArgumentNullException(nameof(modelstate));
             }
 
-            if (metadataProvider == null)
-            {
-                throw new ArgumentNullException(nameof(metadataProvider));
-            }
-
             // If modelkey is empty, we need to iterate through properties (obtained from ModelMetadata) and
             // clear validation state for all entries in ModelStateDictionary that start with each property name.
             // If modelkey is non-empty, clear validation state for all entries in ModelStateDictionary
             // that start with modelKey
             if (string.IsNullOrEmpty(modelKey))
             {
-                var modelMetadata = metadataProvider.GetMetadataForType(modelType);
-                var elementMetadata = modelMetadata.ElementMetadata;
-                if (elementMetadata != null)
+                if (modelMetadata.IsCollectionType)
                 {
-                    modelMetadata = elementMetadata;
+                    foreach (var kvp in modelstate)
+                    {
+                        if (kvp.Key.Length > 0 && kvp.Key[0] == '[')
+                        {
+                            // Starts with an indexer
+                            kvp.Value.Errors.Clear();
+                            kvp.Value.ValidationState = ModelValidationState.Unvalidated;
+                        }
+                    }
                 }
-
-                foreach (var property in modelMetadata.Properties)
```

Comment:
Notice that the existing code isn't really correct either. It only handles collections of POCOs correctly, not collections of simple types.